src/Mono.Android/Android.Runtime/JNIEnvInit.cs

```diff
@@ -22,7 +22,6 @@ internal struct JnienvInitializeArgs {
 			public IntPtr          Class_forName;
 			public uint            logCategories;
 			public int             version; // TODO: remove, not needed anymore
-			public int             androidSdkVersion;
```

Comment:
As per [your comment on Discord](https://discord.com/channels/732297728826277939/732297837953679412/1283472357008347218):

> Ok, I found maybe the size of `JnienvInitializeArgs` in C++ / C# didn't match
I removed from one and not the other

this means we should *also* ***add*** a `public ulong struct_size` field to the beginning of the struct, and assert that the size matches:

```csharp
internal partial struct JnienvInitializeArgs {
    public static along ExpectedStructSize = IntPtr.Size == 8 ? XXX : YYY;
    public ulong struct_size;
    // …
}
```

Update the C++ side to be consistent:

In C++ land, set the new `struct_size` field in [`init_android_runtime()`](https://github.com/dotnet/android/blob/8298b6f7698d92771c558f94d88917e2c6a96b62/src/native/monodroid/monodroid-glue.cc#L817):

```cpp
struct JnienvInitializeArgs init = {};
init.struct_size = sizeof(JnienvInitializeArgs);
```

And finally in `Initialize()`, assert that things match:

```csharp
if (args->struct_size != JnienvInitializeArgs.ExpectedStructSize) {
    FailFast(…);
}
```

While doing this, we may also want to reorder the struct members so that pointers are stored together, to minimize internal alignment padding.

Reply:
`sizeof()` on the native size won't work as expected in this case.  It will include padding of the structure which may differ depending on the target platform (even if bitness is the same).  We'd have to calculate structure size as a sum of individual field sizes (a compile-time calculation), but then the managed size would have to compute it at run time, a bit of time loss.

However, perhaps we can just add a "serial number" field to the structure and bump it every time the structure changes, with comments to update both the managed and native sides at the same time?

Reply:
I added a comment to both `struct`s. I think if I had seen that, I would have caught it earlier.